File: 2022/src/day8/solution.ts
import { readFileSync } from "fs";
import path from "path";

const file = readFileSync(path.join(__dirname, "./input.txt"), "utf-8");
const matrix = file
  .split("\n")
  .map((row) => row.split("").map((x) => Number(x)));

const getColAsRow = (matrix: number[][], col: number) => {
  return matrix.map((row) => row[col]);
};

const isVisible = (matrix: number[][], x: number, y: number) => {
  const row = matrix[y];
  const col = getColAsRow(matrix, x);
  const itemHeight = matrix[y][x];
  // console.log({ itemHeight, x, y });

  const coveredFromLeft = row.slice(0, x).some((item) => item >= itemHeight);
  const coveredFromRight = row
    .slice(x + 1, row.length)
    .some((item) => item >= itemHeight);
  const coveredFromTop = col.slice(0, y).some((item) => item >= itemHeight);
  const coveredFromBottom = col
    .slice(y + 1, col.length)
    .some((item) => item >= itemHeight);

  return [
    coveredFromTop,
    coveredFromRight,
    coveredFromBottom,
    coveredFromLeft,
  ].every((x) => x);
};

const getScenicScore = (matrix: number[][], x: number, y: number) => {
  const row = matrix[y];
  const col = getColAsRow(matrix, x);
  const itemHeight = matrix[y][x];

  if (x === 0 || y === 0 || y === col.length || x === row.length) {
    console.log("edge");
    return 0;
  }
  const indexCoveredLeft = row
    .slice(0, x)
    .reverse()
    .findIndex((x) => x >= itemHeight);
  const indexCoveredRight = row
    .slice(x + 1, row.length)
    .findIndex((x) => x >= itemHeight);

  const indexCoveredTop = col
    .slice(0, y)
    .reverse()
    .findIndex((x) => x >= itemHeight);
  const indexCoveredBottom = col
    .slice(y + 1, col.length)
    .findIndex((x) => x >= itemHeight);

  return (
    (indexCoveredTop === -1 ? y : indexCoveredTop + 1) *
    (indexCoveredRight === -1 ? row.length - x - 1 : indexCoveredRight + 1) *
    (indexCoveredBottom === -1 ? col.length - y - 1 : indexCoveredBottom + 1) *
    (indexCoveredLeft === -1 ? x : indexCoveredLeft + 1)
  );
};

let totalCount = 0;
let scenicScores: number[] = [];

for (let rowIndex = 0; rowIndex < matrix.length; rowIndex++) {
  for (let colIndex = 0; colIndex < matrix[rowIndex].length; colIndex++) {
    const visible = isVisible(matrix, colIndex, rowIndex) ? 0 : 1;
    totalCount += visible;

    const scenicScore = getScenicScore(matrix, colIndex, rowIndex);
    scenicScores.push(scenicScore);
  }
}


console.log(Math.max(...scenicScores));

File: 2022/src/day8/solution.rs
use aoc_2022::read_input;

const DAY: i8 = 8;

fn get_col_as_row(matrix: &Vec<Vec<i32>>, col: usize) -> Vec<i32> {
    return matrix.iter().map(|row| row[col]).collect();
}

fn is_visible(matrix: &Vec<Vec<i32>>, x: usize, y: usize) -> bool {
    let row = &matrix[y];
    let col = get_col_as_row(matrix, x);
    let item_height = matrix[y][x];

    let covered_from_left = row[..x].iter().any(|&item| item >= item_height);
    let covered_from_right = row[x + 1..].iter().any(|&item| item >= item_height);
    let covered_from_top = col[..y].iter().any(|&item| item >= item_height);
    let covered_from_bottom = col[y + 1..].iter().any(|&item| item >= item_height);

    return [
        covered_from_top,
        covered_from_right,
        covered_from_bottom,
        covered_from_left,
    ]
    .iter()
    .all(|&x| x);
}

fn get_scenic_score(matrix: &Vec<Vec<i32>>, x: usize, y: usize) -> usize {
    let row = &matrix[y];
    let col = get_col_as_row(matrix, x);
    let item_height = matrix[y][x];

    if x == 0 || y == 0 || y == col.len() || x == row.len() {
        return 0;
    }

    let index_covered_left = row[..x - 1].iter().rev().position(|&x| x >= item_height);
    let index_covered_right = row[x + 1..].iter().position(|&x| x >= item_height);
    let index_covered_top = col[..y - 1].iter().rev().position(|&x| x >= item_height);
    let index_covered_bottom = col[y + 1..].iter().position(|&x| x >= item_height);

    return (match index_covered_top {
        Some(x) => x + 1,
        None => y,
    } * match index_covered_right {
        Some(x) => row.len() - x - 1,
        None => x,
    } * match index_covered_bottom {
        Some(x) => col.len() - y - 1,
        None => y,
    } * match index_covered_left {
        Some(x) => x + 1,
        None => x,
    });
}

fn main() {
    let input = read_input(DAY);
    let lines = input.lines();
    let matrix: Vec<Vec<i32>> = lines
        .map(|row| {
            row.split("")
                .filter_map(|item| match item {
                    "" => None,
                    _ => Some(item.parse::<i32>().unwrap()),
                })
                .collect()
        })
        .collect();

    let mut total_count: usize = 0;
    let mut scenic_scores: Vec<usize> = Vec::new();

    for row_index in 0..matrix.len() {
        for col_index in 0..matrix[row_index].len() {
            let visible = if is_visible(&matrix, col_index, row_index) {
                0
            } else {
                1
            };
            total_count += visible;

            let scenic_score = get_scenic_score(&matrix, col_index, row_index);
            scenic_scores.push(scenic_score);
        }
    }

    println!("{}", scenic_scores.iter().max().unwrap());
}
